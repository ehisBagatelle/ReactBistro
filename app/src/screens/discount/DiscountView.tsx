import React, { useState } from "react";
import {
  Text,
  View,
  FlatList,
  TouchableOpacity,
  Image,
  StyleSheet,
} from "react-native";
import Layout from "../../components/Layout";
import { Discount } from "../../services/OrderService";

interface Props {
  discounts: Discount[];
  appliedDiscounts: Discount[];
  onPressDiscount: (discount: Discount) => void;
}

const DiscountView: React.FC<Props> = ({
  discounts,
  appliedDiscounts,
  onPressDiscount,
}: Props) => {
  const renderItem = ({ item }: { item: Discount }) => (
    <TouchableOpacity
      onPress={() => onPressDiscount(item)}
      style={styles.discountItem}
    >
      <Text>
        {item.type === "dollar" ? `$${item.amount}` : `${item.amount}%`}
      </Text>
      {appliedDiscounts.find((discount) => discount.id === item.id) && (
        <Image style={styles.icon} source={require("../../../assets/icons8-check-50.png")} />
      )}
    </TouchableOpacity>
  );

  return (
    <Layout>
      <View>
        <FlatList
          data={discounts}
          keyExtractor={(item) => item.id.toString()}
          renderItem={renderItem}
        />
      </View>
    </Layout>
  );
};

const styles = StyleSheet.create({
  discountItem: {
    padding: 16,
    borderBottomWidth: 1,
    borderBottomColor: "#ccc",
    display: "flex",
    justifyContent: "space-between",
    width: "100%",
    flexDirection: "row",
    alignItems: "center",
    height: 50
  },
  icon: {
    height: 30, 
    width: 30
  }
});

export default DiscountView;
