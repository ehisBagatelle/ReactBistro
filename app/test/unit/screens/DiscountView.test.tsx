import React from "react";
import renderer from "react-test-renderer";
import { TouchableOpacity, Image } from "react-native";
import DiscountView from "../../../src/screens/discount/DiscountView";
import { discounts } from "../../fixtures/TestData";

describe("DiscountView component", () => {
  const mockDiscounts = [
    { id: 1, type: "dollar", amount: 5 },
    { id: 2, type: "percentage", amount: 10 },
  ];

  const mockAppliedDiscounts = [
    mockDiscounts[0]
  ];

  const mockOnPressDiscount = jest.fn();

  it("renders correctly with discounts", () => {
    const tree = renderer
      .create(
        <DiscountView
          discounts={discounts}
          appliedDiscounts={[discounts[0]]}
          onPressDiscount={mockOnPressDiscount}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("calls onPressDiscount when a discount is pressed", () => {
    const component = renderer.create(
      <DiscountView
        discounts={[discounts[0]]}
        appliedDiscounts={[]}
        onPressDiscount={mockOnPressDiscount}
      />
    );

    const touchableOpacity = component.root.findByType(TouchableOpacity);
    touchableOpacity.props.onPress();
    expect(mockOnPressDiscount).toHaveBeenCalledWith(discounts[0]);
  });

  it("renders correctly without discounts", () => {
    const tree = renderer
      .create(
        <DiscountView
          discounts={[]}
          appliedDiscounts={[]}
          onPressDiscount={mockOnPressDiscount}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders correctly without applied discounts", () => {
    const tree = renderer
      .create(
        <DiscountView
          discounts={discounts}
          appliedDiscounts={[]}
          onPressDiscount={mockOnPressDiscount}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
