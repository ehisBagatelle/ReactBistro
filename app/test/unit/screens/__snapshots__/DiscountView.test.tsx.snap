// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountView component renders correctly with discounts 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "flex-start",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingVertical": 10,
        "width": "100%",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons8-bistro-64.png",
        }
      }
    />
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 60,
          "fontStyle": "normal",
          "fontWeight": "400",
        }
      }
    >
      React Bistro
    </Text>
  </View>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View>
      <RCTScrollView
        data={
          [
            {
              "amount": 5,
              "id": 3233,
              "type": "dollar",
            },
            {
              "amount": 10,
              "id": 3234,
              "type": "percentage",
            },
            {
              "amount": 20,
              "id": 3235,
              "type": "percentage",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#ccc",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 16,
                  "width": "100%",
                }
              }
            >
              <Text>
                $5
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../assets/icons8-check-50.png",
                  }
                }
                style={
                  {
                    "height": 30,
                    "width": 30,
                  }
                }
              />
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#ccc",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 16,
                  "width": "100%",
                }
              }
            >
              <Text>
                10%
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#ccc",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 16,
                  "width": "100%",
                }
              }
            >
              <Text>
                20%
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;

exports[`DiscountView component renders correctly without applied discounts 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "flex-start",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingVertical": 10,
        "width": "100%",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons8-bistro-64.png",
        }
      }
    />
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 60,
          "fontStyle": "normal",
          "fontWeight": "400",
        }
      }
    >
      React Bistro
    </Text>
  </View>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View>
      <RCTScrollView
        data={
          [
            {
              "amount": 5,
              "id": 3233,
              "type": "dollar",
            },
            {
              "amount": 10,
              "id": 3234,
              "type": "percentage",
            },
            {
              "amount": 20,
              "id": 3235,
              "type": "percentage",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#ccc",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 16,
                  "width": "100%",
                }
              }
            >
              <Text>
                $5
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#ccc",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 16,
                  "width": "100%",
                }
              }
            >
              <Text>
                10%
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "borderBottomColor": "#ccc",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "row",
                  "height": 50,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 16,
                  "width": "100%",
                }
              }
            >
              <Text>
                20%
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;

exports[`DiscountView component renders correctly without discounts 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "flex-start",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "flex-start",
        "paddingVertical": 10,
        "width": "100%",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons8-bistro-64.png",
        }
      }
    />
    <Text
      style={
        {
          "color": "#000",
          "fontSize": 60,
          "fontStyle": "normal",
          "fontWeight": "400",
        }
      }
    >
      React Bistro
    </Text>
  </View>
  <View
    style={
      {
        "width": "100%",
      }
    }
  >
    <View>
      <RCTScrollView
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</View>
`;
